lucene/core/src/java/org/apache/lucene/search/comparators/DocComparator.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search.comparators;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.Scorable;
+
+import java.io.IOException;
+
+/**
+ * Comparator that sorts by asc _doc
+ */
+public class DocComparator extends FieldComparator<Integer> {
+    private final int[] docIDs;
+    private final boolean enableSkipping; // if skipping functionality should be enabled
+    private int bottom;
+    private int topValue;
+    private boolean topValueSet;
+    private boolean bottomValueSet;
+    private boolean hitsThresholdReached;
+
+    /** Creates a new comparator based on document ids for {@code numHits} */
+    public DocComparator(int numHits, boolean reverse) {
+        this.docIDs = new int[numHits];
+        // skipping functionality is enabled if we are sorting by _doc in asc order
```

Comment:
Should it be activated only if the comparator is used as the primary sort ?

Reply:
@jimczi Thanks Jim. We also had an implicit protection for a primary sort in [MultiLeafFieldComparator.java](https://github.com/apache/lucene-solr/pull/1725/files#diff-fbcae9439c314674b6d465e171fe535eR95) i, but I agree it is good to explicitly indicate a primary sort. 

Addressed in 485fe4f

Reply:
> We also had an implicit protection for a primary sort 

Ok I see so the expectation is that `setHitsThresholdReached` is only called on the primary sort and we use it as a signal to enable skipping. 

> Addressed in 485fe4f

I was thinking that we could use the information exposed in [`SortField#getComparator`](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/SortField.java#L448), we don't need to add a new callback. However, I think we can rely on what you describe above, `setHitsThresholdReached` and `competitiveIterator` are the callbacks that we need to enable the optimization. Sorry for the back and forth but I we probably don't need 485fe4f ;).

Reply:
@jimczi Thanks for the feedback.

> I was thinking that we could use the information exposed in SortField#getComparator, we don't need to add a new callback

Good point, I haven't noticed that before, I've refactored the code to use `sortPos` from this function.


> However, I think we can rely on what you describe above, setHitsThresholdReached and competitiveIterator are the callbacks that we need to enable the optimization.   ... we probably don't need 485fe4f

Doing sort optimization only on primary sort in `MultiLeafFieldComparator` is implicit and some refactoring of it may bring wrong result. So I think it is good to have extra protection